pandas/core/reshape/reshape.py

```diff
@@ -743,7 +746,16 @@ def _convert_level_number(level_num: int, columns: Index):
         if slice_len != levsize:
             chunk = this.loc[:, this.columns[loc]]
             chunk.columns = level_vals_nan.take(chunk.columns.codes[-1])
-            value_slice = chunk.reindex(columns=level_vals_used).values
+            # Override fill value to prevent upcasting to float64
+            # if we have lower precision floats
+            common_type = find_common_type(chunk.dtypes.tolist())
+            if common_type.kind == "f" and not is_extension_array_dtype(common_type):
+                fill_value = common_type.type(np.nan)
+            else:
+                fill_value = np.nan
+            value_slice = chunk.reindex(
+                columns=level_vals_used, fill_value=fill_value
+            ).values
```

Comment:
This is something we'll have to figure out at some point, right now this coerces EA dtypes to object. But ok for another pr since we had this problem before as well...